src/Features/LanguageServer/Protocol/LanguageServerProtocol.cs

```diff
@@ -1,45 +1,161 @@
 
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Composition;
 using System.Threading;
 using System.Threading.Tasks;
-using Microsoft.CodeAnalysis.Editor;
+using Microsoft.CodeAnalysis.LanguageServer.CustomProtocol;
+using Microsoft.CodeAnalysis.LanguageServer.Handler;
 using LSP = Microsoft.VisualStudio.LanguageServer.Protocol;
 
 namespace Microsoft.CodeAnalysis.LanguageServer
 {
     /// <summary>
     /// Implements Language Server Protocol
+    /// TODO - Make this public when we're ready.
     /// </summary>
-    public sealed class LanguageServerProtocol
+    [Shared]
+    [Export(typeof(LanguageServerProtocol))]
+    internal sealed class LanguageServerProtocol
     {
-        // for now, it is null but later we might use this to provide info on types on dlls
-        private readonly IMetadataAsSourceFileService _metadataAsSourceService = null;
+        private readonly ImmutableDictionary<string, Lazy<IRequestHandler, IRequestHandlerMetadata>> _requestHandlers;
 
-        // TODO - Move hierarchicalDocumentSymbolSupport to client capabilities.
-        // https://github.com/dotnet/roslyn/projects/45#card-20033973
-        private readonly bool _hierarchicalDocumentSymbolSupport;
-        private readonly LSP.ClientCapabilities _clientCapabilities;
+        [ImportingConstructor]
+        public LanguageServerProtocol([ImportMany] IEnumerable<Lazy<IRequestHandler, IRequestHandlerMetadata>> requestHandlers)
+        {
+            _requestHandlers = CreateMethodToHandlerMap(requestHandlers);
+        }
+
+        private ImmutableDictionary<string, Lazy<IRequestHandler, IRequestHandlerMetadata>> CreateMethodToHandlerMap(IEnumerable<Lazy<IRequestHandler, IRequestHandlerMetadata>> requestHandlers)
+        {
+            var requestHandlerDictionary = new Dictionary<string, Lazy<IRequestHandler, IRequestHandlerMetadata>>();
+            foreach (var lazyHandler in requestHandlers)
+            {
+                requestHandlerDictionary.Add(lazyHandler.Metadata.MethodName, lazyHandler);
+            }
+
+            return requestHandlerDictionary.ToImmutableDictionary();
+        }
 
-        public LanguageServerProtocol(LSP.ClientCapabilities clientCapabilities, bool hierarchicalDocumentSymbolSupport)
+        private async Task<ResponseType> ExecuteRequestAsync<RequestType, ResponseType>(string methodName, Solution solution, RequestType request,
+            LSP.ClientCapabilities clientCapabilities, CancellationToken cancellationToken)
         {
-            _hierarchicalDocumentSymbolSupport = hierarchicalDocumentSymbolSupport;
-            _clientCapabilities = clientCapabilities;
+            var handler = (IRequestHandler<RequestType, ResponseType>)_requestHandlers[methodName].Value;
+            return await handler.HandleRequestAsync(solution, request, clientCapabilities, cancellationToken).ConfigureAwait(false);
```

Comment:
looks like you don't need await here?

Reply:
I'm not sure what you mean here, HandleRequestAsync is an async operation

Reply:
but there is no code after await so. you get same result with

```csharp
private Task<ResponseType> ExecuteRequestAsync<RequestType, ResponseType>(string methodName, Solution solution, RequestType request, LSP.ClientCapabilities clientCapabilities, CancellationToken cancellationToken)
{
	var handler = (IRequestHandler<RequestType, ResponseType>)_requestHandlers[methodName].Value;
	return handler.HandleRequestAsync(solution, request, clientCapabilities, cancellationToken);
}
```